File: Python3/aoc24/day02/solve.py
"""Module containing computation of solutions."""

MIN_DELTA: int = 1
MAX_DELTA: int = 3


def solve_part_one(reports: list[list[int]]) -> int:
    """Compute solution of part one."""
    return sum(is_safe(report) for report in reports)


def is_safe(report: list[int]) -> bool:
    """Determine if a report is safe."""
    is_increasing: bool = report[1] >= report[0]

    for index in range(1, len(report)):
        variation: bool = report[index] >= report[index - 1]
        if variation != is_increasing:
            return False

        delta: int = abs(report[index] - report[index - 1])
        if not (MIN_DELTA <= delta <= MAX_DELTA):
            return False

    return True


def solve_part_two(reports: list[list[int]]) -> int:
    """Compute solution of part two."""
    return sum(is_safe(report) or is_pseudo_safe(report) for report in reports)


def is_pseudo_safe(report: list[int]) -> bool:
    """Determine if a report is pseudo-safe.

    I.e. if the report is safe with an element removed.

    """
    for index in range(len(report)):
        subreport: list[int] = list(report)  # Clone
        subreport.pop(index)
        if is_safe(subreport):
            return True

    return False

File: Rust/src/day02/solve.rs
const MIN_DELTA: i32 = 1;
const MAX_DELTA: i32 = 3;

pub fn solve_part_one(reports: &[Vec<i32>]) -> i128 {
    reports
        .iter()
        .map(|report| match is_safe(report) {
            true => 1,
            false => 0,
        })
        .sum::<i128>()
}

fn is_safe(report: &[i32]) -> bool {
    let is_increasing: bool = report.get(1).unwrap() >= report.first().unwrap();

    for index in 1..report.len() {
        let val1: &i32 = report.get(index - 1).unwrap();
        let val2: &i32 = report.get(index).unwrap();

        if (val2 >= val1) != is_increasing {
            return false;
        }

        let delta: i32 = (val2 - val1).abs();
        if !(MIN_DELTA..=MAX_DELTA).contains(&delta) {
            return false;
        }
    }

    true
}

pub fn solve_part_two(reports: &[Vec<i32>]) -> i128 {
    reports
        .iter()
        .map(|report| match is_safe(report) || is_pseudo_safe(report) {
            true => 1,
            false => 0,
        })
        .sum::<i128>()
}

// Brute-force sub-report safety check, cost is low
fn is_pseudo_safe(report: &[i32]) -> bool {
    for index in 0..report.len() {
        let mut subreport: Vec<i32> = report.to_owned();
        subreport.remove(index);

        if is_safe(&subreport) {
            return true;
        }
    }
    false
}
